Fix #5: GAE Abstraction part 1 - Topic Page handler [Blocked: #85]

rt4914 opened the pull request:
## Explanation
The PR contains data module, necessary libraries for API calls and Topic Management System models, retrofit instance and interceptor.

Also this PR was earlier on #61 and #75 

## Libraries Used
1. **Retrofit:** implementation 'com.squareup.retrofit2:retrofit:2.5.0'
This library is responsible for making API calls to Oppia backend
2. **Retrofit Moshi Converter:** implementation 'com.squareup.retrofit2:converter-moshi:2.5.0'
This library is the actual deserializer that we are using over GSON.
3. **Moshi Kotlin Support:** implementation"com.squareup.moshi:moshi-kotlin:1.8.0"
Moshi Kotlin support library for using Moshi in Kotlin.

## Checklist
Work items:

- [x] Identify among all existing GAE endpoints (get/post), which we can use during the prototype
- [x] Loosely identify data missing from the above endpoints (this bucket is expected to be small/non-existent), and identify which endpoints have data which needs to be changed for the Android app
- [x] Specify the Retrofit interfaces to represent all of the identified endpoints (simple interfaces/no annotations or return types needed beyond pseudocode)
- [x] Identify** what data (via pseudocode data structures/YAML/JSON/etc) we will be passing along from the backend through these endpoints
- [ ] Identify how error handling will work (e.g. what Retrofit does), whether RPC retry is supported & how it is/can be configured, etc.
- [x] Determine the testing strategy for the GAE endpoints for downstream app components

BenHenning committed:
Initial introduction of the data source module.

High-level notes:
- Data sources are now being referred to as 'DataProviders'
- This commit also fixes the build which was broken in a previous PR
  ('compile' needs to be used for the proto library)
- The data source module is referred to as just 'data'
- This introduces a few style changes to prohibit wildcard imports
- This also normalizes the JDK versions needed across modules, potentially
  fixing some of the repeated reversions the IDE does for a Java SDK
  version setting

Functional differences: the app now properly shows the correct string upon
initial open ("Welcome to Oppia"), and reopening thereafter correctly
shows the updated string ("Welcome back to Oppia"). Rotating the device is
currently broken without Dagger integration.

Overall, this moves all general data provider functionality into the new
module, and introduces a PersistentCacheStore to store proto messages
on-disk, or load them into memory if they are available. Overall, this
commit is introducing a potentially reasonable medium-term solution for
data providers and on-disk proto storage. A long-term design should still
be determined, but this seems like a good place to start.

Tests are currently broken or missing. New TODOs will also be resolved in
a later commit.

Commit:
Data module dependencies, retrofit and models

Commit:
Removed kapt from build.gradle in data

Commit:
Javadoc comments

rt4914 commented:
@BenHenning PTAL. This PR should help @veena14cs in topic-index-handler as well as I will use this branch for my future PRs for other APIs.

BenHenning commented on data/src/main/java/org/oppia/data/Constants.kt:
```diff
@@ -0,0 +1,8 @@
+/** Constant values for data module will be defined here */
+object Constants {
+
+  const val RESPONSE_SUCCESS = 200
```
Unless this is going to be processed outside of this file, we could just add it as a module-level constant to the interceptor.

Please also add a Javadoc to it, or name it something self documenting like HTTP_OK.

rt4914 replied:
Unable to understand your first line. Do you mean I should declare this constant in gradle.properties ?

Also, I have renamed RESPONSE_SUCCESS oto HTTP_OK

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/NetworkInterceptor.kt:
```diff
@@ -0,0 +1,42 @@
+   * @param rawJson: This is the string that we get in body of our response
+   * @return String: rawJson without XSSI_PREFIX
+   */
+  fun removeXSSIPrefixFromResponse(rawJson: String): String {
```
Since this is a helper method, let's make it private since it shouldn't be accessed outside of this class.

When testing the interceptor, we should only test it from its public API to have parity with how other classes will interact with it. In this case, we should use intercept() to test that the XSSI prefix is properly removed.

Is there a way to use Retrofit in tests? https://riggaroo.co.za/retrofit-2-mocking-http-responses/ suggests that this is possible.

This is actually a great opportunity to demonstrate how to mock HTTP responses in Retrofit, since that will be the foundation for our later downstream fakes that we use to support the HTTP side of the entire app infrastructure for UI tests.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/NetworkInterceptor.kt:
```diff
@@ -0,0 +1,42 @@
+
+/**
+ * Interceptor on top of Retrofit to modify requests and response
+ * The Interceptor intercepts requests and response and in every response
```
Please wrap Javadoc comments to the 120 character limit, or put new lines between Javadoc paragraphs.

rt4914 replied:
@BenHenning PTAL.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/OppiaGaeClient.kt:
```diff
@@ -0,0 +1,26 @@
+        .client(client.build())
+        .build()
+    }
+    return@lazy retrofit
```
Out of curiosity, do we actually need to use ``return@lazy`` here? If so, why? I haven't needed that yet for lazy initialization. Do you know what it does?

rt4914 replied:
Actually, in my case it was giving error if I use `return retrofit` and the solution for that was this.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/OppiaGaeClient.kt:
```diff
@@ -0,0 +1,26 @@
+  private var retrofit: Retrofit? = null
+
+  val retrofitInstance: Retrofit? by lazy {
+    if (retrofit == null) {
```
Is this additional logic needed? Can't we just initialize Retrofit lazily? I don't completely follow why we need this dual lazy logic.

rt4914 replied:
Check the updated code for this.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/OppiaGaeClient.kt:
```diff
@@ -0,0 +1,26 @@
+import retrofit2.Retrofit
+import retrofit2.converter.moshi.MoshiConverterFactory
+
+/** This file manages Retrofit configuration and connection with Oppia backend */
```
Since this is an object, we should phrase this documentation based on what the object contains. Consider, e.g.:

``/** An object that contains the Retrofit configuration with the Oppia backend. */``

rt4914 replied:
Done

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt:
```diff
@@ -0,0 +1,10 @@
+object NetworkSettings {
+
+  // TODO(#74): Move this to DI graph
+  const val BASE_URL = "https://oppia.org"
```
I suggest we add a DEVELOPER_URL that points to localhost (similar to the demo app), switch this to be a PROD_URL, and have a BASE_URL which points to DEVELOPER_URL by default (requiring developers to set up a local instance and use that).

Please add Javadocs for each of these URLs explaining what they represent (especially BASE_URL since it's used directly by the Retrofit configuration).

rt4914 replied:
I am multiple doubts in this.
1. DEVELOPER_URL similar to demo app? I haven't seen that anywhere. 
2. For switching I am introduing a boolean value which will switch between developer and prod server.

rt4914 replied:
I have removed BASE_URL and intoduced a `fun getBaseUrl()` and using boolean value this method switches between the server.

rt4914 replied:
Also, not sure what should be the exact value of PROD and DEVELOPER URL for now.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt:
```diff
@@ -0,0 +1,10 @@
+
+  // TODO(#74): Move this to DI graph
+  const val BASE_URL = "https://oppia.org"
+  const val XSSI_PREFIX = ")]}\'"
```
Please add documentation for where this comes from & why we need it. If possible, it would be nice to point to where in the Oppia backend this is used, as well.

rt4914 replied:
Added github link for XSSI_PREFIX along with other usage details.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/api/TopicService.kt:
```diff
@@ -0,0 +1,14 @@
+interface TopicService {
+
+  @GET("topic_data_handler/{topic_name}")
+  fun getTopicByTopicName(@Path("topic_name") topicName: String): Call<Topic>
```
Nit: ``getTopicByName`` is probably sufficient, and it's a bit more succinct.

rt4914 replied:
Done

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/api/TopicService.kt:
```diff
@@ -0,0 +1,14 @@
+import retrofit2.http.GET
+import retrofit2.http.Path
+
+/** Retrofit instance for managing topic-handler */
```
Maybe just: ``Service that provides access to topic endpoints.``

rt4914 replied:
Done

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt:
```diff
@@ -0,0 +1,24 @@
+
+/**
+ * Model for serialization/deserialization using Moshi with Retrofit *
+ * @see <a href="https://github.com/oppia/oppia/blob/b33aa9cf1aa6372e12d0b35f95cceb44efe5320f/core/controllers/topic_viewer.py#L45">Oppia-Web Topic structure</a>
```
Is this URL pointing to the right thing? It seems to point to a topic data handler rather than a to_dict method.

rt4914 replied:
I think it is pointing correctly because the response is a mixture of multiple responses from different function calls.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/StorySummary.kt:
```diff
@@ -0,0 +1,17 @@
+
+/**
+ * Model for serialization/deserialization using Moshi with Retrofit
+ * @see <a href="https://github.com/oppia/oppia/blob/b33aa9cf1aa6372e12d0b35f95cceb44efe5320f/core/domain/story_domain.py#L1118">Oppia-Web StorySummary structure</a>
```
Two things regarding these links:

1. Since Kotlin uses markdown for Javadocs, you can make this a bit shorter by avoiding the href anchor tag
2. You can shorten these blob links by just using the first 6 hash characters, e.g. https://github.com/oppia/oppia/blob/b33aa9/core/domain/story_domain.py#L1118.

Ditto elsewhere in this PR.

rt4914 replied:
1. I tried this `@see [link](https://github.com/oppia/oppia/blob/b33aa9/core/domain/story_domain.py#L1118)` , but it was not working. Can you guide me this?
2. Changed blob links to 6 hash characters only.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/StorySummary.kt:
```diff
@@ -0,0 +1,17 @@
+import com.squareup.moshi.JsonClass
+
+/**
+ * Model for serialization/deserialization using Moshi with Retrofit
```
Let's describe what this model actually is, rather than how it should be used. Ditto elsewhere in this PR.

rt4914 replied:
Changed the description for all models

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/StorySummary.kt:
```diff
@@ -0,0 +1,17 @@
+ * @see <a href="https://github.com/oppia/oppia/blob/b33aa9cf1aa6372e12d0b35f95cceb44efe5320f/core/domain/story_domain.py#L1118">Oppia-Web StorySummary structure</a>
+ */
+@JsonClass(generateAdapter = true)
+data class StorySummary(
```
I believe at some point we discussed suffixing all of these GAE models with 'Gae' to help avoid conflicts with future domain models that these will be converted through. Is that still the case? If so, let's go ahead and make that change to these model classes.

rt4914 replied:
Added `Gae` prefix to all models

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt:
```diff
@@ -0,0 +1,24 @@
+  @Json(name = "topic_name") val topicName: String?,
+  @Json(name = "canonical_story_dicts") val canonicalStoryDicts: List<StorySummary?>?,
+  @Json(name = "additional_story_dicts") val additionalStoryDicts: List<StorySummary?>?,
+  /** skill_descriptions map has skill id as key and skill name as value */
```
Consider phrasing this instead as something like:

``/** A map of skill descriptions keyed by skill ID. */``

That way the documentation is describing the property, and still includes the important bit about how the map is keyed & what it contains.

Ditto for ``degreesOfMastery`` below.

rt4914 replied:
Done.

BenHenning commented on data/src/test/java/org/oppia/data/NetworkInterceptorTest.kt:
```diff
@@ -0,0 +1,33 @@
+class NetworkInterceptorTest {
+
+  @Test
+  fun testNetworkInterceptor_removeXSSIPrefixFromResponse_withXSSIPREFIX() {
```
Nit on naming: when camel casing with fully capitalized names, you can lowercase all the letters after the first one (see https://google.github.io/styleguide/javaguide.html#s5.3-camel-case). In this case:

``testNetworkInterceptor_removeXssiPrefixFromResponse_withXssiPrefix``

Also, test names can be clearer when phrased as:

*testAction_withOneCondition_andOtherCondition_hasExpectedOutcome*

Your name is close to this, but consider the alternative:

``testNetworkInterceptor_withXssiPrefix_removesXssiPrefix``

If all tests are phrased this way, it will be really easy to see which behaviors are being tested which can help identify missing cases for more complex classes.

rt4914 replied:
@BenHenning Thanks a lot. This makes sense.

BenHenning reviewed (changes requested):
Thanks @rt4914! I did a full review, though I'm expecting the tests to change some per my comment about testing via public APIs.

Generally, the PR looks really good! I think one more thorough pass will be sufficient to get this checked in since I'm especially interested in seeing tests that we can use as canonical examples for all future HTTP test code. Please let me know if you have any questions, concerns, or doubts about any of my comments or the PR.

BenHenning committed:
Introduce one possible integration with Dagger 2.

This is loosely based on both
https://github.com/tfcporciuncula/dagger-simple-way and
https://dagger.dev/android.html, except this approach is meant to keep
activities and fragments as thin as possible and push as much UI
presentation and service business logic into Dagger-provided objects as
possible. This maximizes Dagger use throughout the codebase, and opens the
potential for future code generation for these thin adapter fragments &
services.

This fixes the rotation bug in the app: rotating upon the initial app open
retains the 'Welcome to Oppia' label since the same user app history
controller is used in both instances of HomeFragment due to it being
singleton scoped and not needing to be recreated.

No tests were updated with these changes.

Commit:
Javadoc comments added

Commit:
Reolved Javadoc mistakes

Commit:
Update build.gradle proto directive to use 'compile' instead of (#81)

Commit:
Merge remote-tracking branch 'upstream/introduce-data-module' into gae-abstraction-base

Commit:
Renaming test cases

rt4914 reviewed (commented):
@BenHenning I have replied to most of your comments and made changes accordingly. I will now work on Retrofit Test cases

Commit:
Test cases for TopicService

Commit:
Made changes to OppiaGaeClient

Commit:
Made changes to OppiaGaeClient

rt4914 commented:
@BenHenning PTAL

BenHenning commented:
NB: This is blocked on #85.

Commit:
TopicService test code using Json response

Commit:
Revert changes in codeStyles/Project.xml

Commit:
Revert changes in codeStyles/Project.xml

rt4914 commented:
@BenHenning PTAL

I have modified the MockTopicService test file such that it takes data from Json directly. Therefore now we can see that Moshi is working as intended but there is still one issue. NetworkInterceptor is not getting used in these test cases.
I have not be able to find the correct solution for API testing on the web. I think we might need to introduce FakeInterceptor but again that would not be the best approach because it is somewhat brute-force testing in a way that we set data to our model in one file and check those values in other file and also FakeInterceptors do not use Json.

Commit:
Optimise imports

Commit:
Check NetworkInterceptor part

rt4914 commented:
> @BenHenning PTAL
> 
> I have modified the MockTopicService test file such that it takes data from Json directly. Therefore now we can see that Moshi is working as intended but there is still one issue. NetworkInterceptor is not getting used in these test cases.
> I have not be able to find the correct solution for API testing on the web. I think we might need to introduce FakeInterceptor but again that would not be the best approach because it is somewhat brute-force testing in a way that we set data to our model in one file and check those values in other file and also FakeInterceptors do not use Json.

I have updated the code in such a way the `removePrefix` method from `NetworkInterceptor` is getting tested. This works for now but it is incorrect approach.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/OppiaGaeClient.kt:
```diff
@@ -0,0 +1,20 @@
+import retrofit2.converter.moshi.MoshiConverterFactory
+
+/** An object that contains the Retrofit configuration with the Oppia backend. */
+object OppiaGaeClient {
```
Let's go ahead and move this to the Dagger graph so that we can inject TopicService directly. We probably don't need OppiaGaeClient at all, and we can also hide the retrofit instance. A pseudocode implementation of this is:

```kotlin
@Module
class NetworkModule {
  @Qualifier private annotation class OppiaRetrofit

  @OppiaRetrofit
  @Provides
  @Singleton
  fun provideRetrofitInstance() {
    return retrofit2.Retrofit.Builder().
      baseUrl(...)
      ...
      .build()
  }

  @Provides
  @Singleton
  fun provideTopicService(@OppiaRetrofit retrofit: Retrofit) {
    return retrofit.create(TopicService::class.java)
  }
}
```

NB: This module needs to be included in the application component's module list.

And with this module, we can now inject ``TopicService`` wherever we need it.

BenHenning commented on data/src/test/java/org/oppia/data/MockTopicService.kt:
```diff
@@ -0,0 +1,33 @@
+    return delegate.returningResponse(topic).getTopicByName(topicName)
+  }
+
+  private fun MockGaeTopic(): GaeTopic {
```
Nit: since this is a method, its name should be lowerCamelCase and a verb phrase, e.g. ``createMockGaeTopic``.

rt4914 replied:
Done.

BenHenning commented on data/src/test/java/org/oppia/data/MockTopicTest.kt:
```diff
@@ -0,0 +1,45 @@
+  private var retrofit: Retrofit? = null
+
+  @Before
+  @Throws(Exception::class)
```
Is this necessary? I thought this is implied in Kotlin since it doesn't require using checked exceptions. I thought this only serves to set Kotlin methods up for Java use, but in this case I believe JUnit methods are called reflectively and are assumed to throw exceptions.

rt4914 replied:
I did search this on internet but haven't found a proper answer for this. I have seen different resources which uses this and there are some resources which do not use this. 

Even I am not sure, whether we actually need this or not.

BenHenning commented on data/src/test/java/org/oppia/data/MockTopicTest.kt:
```diff
@@ -0,0 +1,45 @@
+
+  @Test
+  @Throws(Exception::class)
+  fun testMockTopicService_withTopicName_success() {
```
Nice test! Great to see this working across the pipeline.

rt4914 replied:
Thanks.

BenHenning reviewed (approved):
Thanks @rt4914--this LGTM. However, if you'd like me to take another look after you integrate with Dagger feel free to re-assign the PR to me.

veena14cs committed:
Fix part of #9: Introduce Retrofit classroom data handler service (#83)

* introduced topic index handler

* Call retrofit data handlers from data module

* Update topic_index_items.xml

* removed extra space

* fixed issues

* Update AndroidManifest.xml

* Removed Ui part. http handlers are included in data module

* done changes in handler and data as per the specs

* Update TopicSummarytDicts.kt

* Update NetworkSettings.kt

* added classroom datahandler

* renamed corresponding classes to classroom

* fixed issues

* Update GaeClassroom.kt

* removed spaces

* Update AndroidManifest.xml

* removed unwanted spaces

* added space

Commit:
Resolve typo

Commit:
Merge branch 'gae-abstraction-base' of https://github.com/oppia/oppia-android into gae-abstraction-base

Commit:
Resolve merge conflicts with develop branch

Commit:
NetworkModule introduced

Commit:
Daggerify NetworkInterceptor

rt4914 commented on data/src/main/java/org/oppia/data/Constants.kt:
```diff
@@ -0,0 +1,9 @@
+
+  /** Constant which defines successful API call */
+  const val HTTP_OK = 200
+
```
@BenHenning Not sure how to dagger-ify this.

rt4914 commented on data/src/main/java/org/oppia/data/backends/gae/NetworkInterceptor.kt:
```diff
@@ -0,0 +1,45 @@
+ * The Interceptor removes XSSI_PREFIX from every response to produce valid Json
+ */
+@Singleton
+class NetworkInterceptor @Inject constructor() : Interceptor {
```
@BenHenning Check this `@Inject constructor()` is this correct?

BenHenning replied:
Yup, this looks perfect!

rt4914 replied:
Thanks

rt4914 replied:
Thanks

rt4914 commented on data/src/main/java/org/oppia/data/backends/gae/NetworkModule.kt:
```diff
@@ -0,0 +1,68 @@
+import javax.inject.Singleton
+
+/**
+ * Sample resource https://github.com/gahfy/Feed-Me/tree/unitTests
```
@BenHenning found this wonderful resource which uses dagger+retrofit and follows MVVP clean code. The test cases written in this can be the best approach for testing `data` module. The only concern is that it might bring lot of changes to the code. Check this link whenever you have time, it is worth going through this code.

BenHenning replied:
Thanks--do you want to present this in one of our upcoming mobile meetings? I wanted to cover testing & Dagger directly, but this might be a nice add-on topic.

rt4914 replied:
I do think it is worth sharing with everyone. I am just not sure about finish this by Monday. Anyways, I will prepare for this and will present on Monday.

rt4914 commented on data/src/main/java/org/oppia/data/backends/gae/NetworkModule.kt:
```diff
@@ -0,0 +1,68 @@
+   */
+  @OppiaRetrofit
+  @Provides
+  @Singleton
```
We might want to use `@Reusable` instead of `@Singleton` only from performance point of you.

Check this link: https://stackoverflow.com/questions/39136042/dagger-reusable-scope-vs-singleton

BenHenning replied:
I actually have doubts about Reusable. I suspect it doesn't provide worthwhile performance gains, and it's a bit scary. Dagger doesn't define well when resuable actually reuses an object versus creating a new one. I think the determinism of having clean separations between object lifetimes and the lack of non-deterministic bugs that arise from that scenario would outweigh any expectantly negligible gains we'd get from reusable.

Happy to discuss further. :)

rt4914 replied:
Okay, I think that makes sense.

Commit:
Make retrofit dependency as api

rt4914 commented on app/build.gradle:
```diff
@@ -90,6 +90,7 @@ dependencies {
   kaptAndroidTest(
       'com.google.dagger:dagger-compiler:2.24'
   )
+  api project(':data')
```
Had to introduce this because of error in access to retrofit
Reference link for solution: https://stackoverflow.com/questions/47124987/dagger-2-cannot-access-retrofit

BenHenning replied:
Does this need to be ``api``? If so, it's a good place to put a TODO on #59 to remove this once we can use Bazel, since ``api`` is deprecated in Gradle and so this solution may eventually break (whereas with Bazel it shouldn't).

rt4914 replied:
Yes we need this `api`. Added TODO accordingly.

rt4914 commented on data/build.gradle:
```diff
@@ -43,20 +44,33 @@ dependencies {
+      'com.squareup.retrofit2:converter-moshi:2.5.0',
+      'com.squareup.retrofit2:retrofit:2.5.0',
+      'com.squareup.retrofit2:retrofit-mock:2.5.0',
+  )
```
Had to introduce this because of error in access to retrofit
Reference link for solution: https://stackoverflow.com/questions/47124987/dagger-2-cannot-access-retrofit

BenHenning replied:
Ditto here re: TODO to convert this a Bazel-based solution.

rt4914 replied:
Done. Reference to #59

Commit:
Javadocs and linking of oppia-web to models

rt4914 commented on data/src/test/java/org/oppia/data/backends/test/NetworkInterceptorTest.kt:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    fun getNetworkInterceptorTest(): NetworkInterceptorTest
+  }
```
@BenHenning Checkout this code for component.

BenHenning replied:
Instead of this & constructing NetworkInterceptor, you want to have an @Inject field for NetworkInterceptor, have a method here to inject the NetworkInterceptorTest, and call that after creating the component in the test's setUp method. For example, see:

https://github.com/oppia/oppia-android/blob/develop/data/src/test/java/org/oppia/data/persistence/PersistentCacheStoreTest.kt.

rt4914 replied:
Updated code accordingly.

BenHenning replied:
This is close, except I think you don't need the Retrofit instance provider (or at least, it's not obvious to me that you do since nothing here seems to use it or any dependencies that need it). You also should inject NetworkInterceptor instead of construct it by using a field, e.g.:

``@Inject lateinit var networkInterceptor: NetworkInterceptor``

This field will be recreated for each test since it'll be reassigned during the set up of the test application component.

rt4914 replied:
Removed Retrofit instance provider and reformatted code using 
`@Inject lateinit var networkInterceptor: NetworkInterceptor`

rt4914 reviewed (commented):
@BenHenning PTAL

BenHenning commented on data/build.gradle:
```diff
@@ -43,20 +44,33 @@ dependencies {
+  api(
+      'com.squareup.retrofit2:converter-moshi:2.5.0',
+      'com.squareup.retrofit2:retrofit:2.5.0',
+      'com.squareup.retrofit2:retrofit-mock:2.5.0',
```
We shouldn't take a prod dependency on retrofit-mock. Can this be test-only, instead? If not, that's a bit unfortunate, and we should have a separate TODO with a new issue to isolate this to a test dependency to avoid pulling it into production builds.

rt4914 replied:
Cannot remove `retrofit-mock` from `api` because we cannot use the mock-retrofit in Test cases and also mentioned TODO linked to #59 and #97

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/NetworkModule.kt:
```diff
@@ -0,0 +1,67 @@
+import javax.inject.Singleton
+
+/**
+ * Sample resource https://github.com/gahfy/Feed-Me/tree/unitTests
```
Nit: let's just make this a separate paragraph in the below Javadoc. Currently, this is a floating Javadoc with no associated element. :)

BenHenning replied:
Actually, I'm not sure where this Javadoc should go since this module isn't related to testing directly. Maybe nowhere for now?

rt4914 replied:
I have shifted to this javadoc because this code has been referenced from that link and therefore it is worth mentioning it here for now. And later we can remove this.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/NetworkModule.kt:
```diff
@@ -0,0 +1,67 @@
+ */
+@Module
+// Safe here as we are dealing with a Dagger 2 module
+@Suppress("unused")
```
This actually isn't needed since the warning is incorrect. The module is being used (and Android Studio will stop complaining about the lack of use after build files are updated).

rt4914 replied:
Removed. @Suppress

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/GaeTopicSummarytDict.kt:
```diff
@@ -0,0 +1,24 @@
+ * https://github.com/oppia/oppia/blob/77d6fd/core/domain/topic_domain.py#L1178
+ */
+@JsonClass(generateAdapter = true)
+data class GaeTopicSummarytDict(
```
This isn't actually a dictionary, it's an object. Is 'GaeTopicSummary' suitable here?

NB: This is a bit nitpicky, but 'dict' is a python structure and should stay probably isolated to python contexts since it's a bit confusing form the Java perspective.

rt4914 replied:
Okay got it. Will change accordingly and also inform will @veena14cs about this.

BenHenning commented on data/src/test/java/org/oppia/data/backends/test/MockTopicTest.kt:
```diff
@@ -0,0 +1,57 @@
+  private lateinit var retrofit: Retrofit
+
+  @Before
+  @Throws(Exception::class)
```
Nit here & below: not needed since these are called reflectively by JUnit so the Java runtime annotation isn't necessary.

rt4914 replied:
Solved. Removed `@Throws`

BenHenning commented on data/src/test/java/org/oppia/data/backends/test/MockTopicTest.kt:
```diff
@@ -0,0 +1,57 @@
+
+  @Test
+  @Throws(Exception::class)
+  fun testMockTopicService_withTopicName_success() {
```
Per the name, this doesn't seem to include the actual action that's being tested. Can you update it to include the operation itself?

Also, it's often more useful to be specific about what's successful (e.g. that the response has the expected topic name). That way we can easily check multiple tests with a similar naming style for thoroughness of behaviors being tested.

rt4914 replied:
Changed this name to `testTopicService_usingFakeJson_deserializationSuccessful` but still not convinced. I am finding it difficult to name the methods correctly. Will need to see multiple code samples to get proper understanding.

BenHenning replied:
This seems reasonable, though you could probably make it slightly clearer by including what is successful, e.g.:

``testService_withFakeJson_deserializesCorrectTopicName``

BenHenning reviewed (approved):
Generally LGTM. Please re-assign to me if you'd like me to do another pass, but as long as you address my comments feel free to merge this.

Commit:
Changes in Javadoc, naming and test cases code

rt4914 commented:
@BenHenning PTAL

BenHenning commented on app/build.gradle:
```diff
@@ -90,6 +90,8 @@ dependencies {
   kaptAndroidTest(
       'com.google.dagger:dagger-compiler:2.24'
   )
+  //TODO (#59): Remove this once Bazel is introduced
```
Nit: space before TODO.

Ditto elsewhere in this PR.

rt4914 replied:
Done.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/NetworkModule.kt:
```diff
@@ -0,0 +1,63 @@
+  private annotation class OppiaRetrofit
+
+  /**
+   * Provides the Retrofit object.
```
Optional nit: we probably don't need Javadocs for module methods since nobody is calling them directly. It's useful in cases with specifically complicated dependencies (e.g. if the injection site has to have some other configuration set up for the injection to succeed).

rt4914 replied:
For now I am keeping this Javadoc.

BenHenning reviewed (approved):
Thanks @rt4914. The recent changes LGTM; left a few comments to improve clarity on a few things.

Commit:
Merge remote-tracking branch 'upstream/develop' into gae-abstraction-base

Commit:
Updated TODO comments and NetworkInterceptorTest

rt4914 commented:
This PR is complete but it does need some more test cases to be introduced which will be finished with other PRs related to services.
